source/features/show-whitespaces.tsx

```diff
@@ -0,0 +1,96 @@
+import './show-whitespaces.css';
+import React from 'dom-chef';
+import select from 'select-dom';
+import features from '../libs/features';
+import onPrFileLoad from '../libs/on-pr-file-load';
+import onNewComments from '../libs/on-new-comments';
+import getTextNodes from '../libs/get-text-nodes';
+
+function showWhiteSpacesOn(line: Element): void {
+	const textNodes = getTextNodes(line);
+
+	for (const textNode of textNodes) {
+		const nodeValue = textNode.textContent!;
+		if (nodeValue.length === 0) {
+			continue;
+		}
+
+		const fragment = document.createDocumentFragment();
+
+		let lastEncounteredCharType;
+		let charType: 'space' | 'tab' | 'other';
+		let node;
+
+		for (const char of nodeValue) {
+			if (char === ' ') {
+				charType = 'space';
+			} else if (char === '\t') {
+				charType = 'tab';
+			} else {
+				charType = 'other';
+			}
+
+			if ((lastEncounteredCharType && lastEncounteredCharType === charType) && node) {
+				node.textContent += char;
+
+				if (charType === 'space') {
+					node.dataset.rghSpaces += '·';
+				} else if (charType === 'tab') {
+					node.dataset.rghTabs += '→';
+				}
+			} else {
+				if (node) {
+					fragment.append(node);
+				}
+
+				if (charType === 'space') {
+					node = <span className="pl-ws pl-space" data-rgh-spaces="·">{char}</span>;
+				} else if (charType === 'tab') {
+					node = <span className="pl-ws pl-tab" data-rgh-tabs="→">{char}</span>;
+				} else {
+					node = <>{char}</>;
+				}
+			}
+
+			lastEncounteredCharType = charType;
+		}
+
+		if (node) {
+			fragment.append(node);
+		}
+
+		textNode.replaceWith(fragment);
+	}
+}
+
+function run(): void {
+	const tables = select.all([
+		'table.js-file-line-container:not(.rgh-showing-whitespace)', // Single blob file, and gist
+		'.file table.diff-table:not(.rgh-showing-whitespace)', // Split and unified diffs
+		'.file table.d-table:not(.rgh-showing-whitespace)' // "Suggested changes" in PRs
+	].join());
+
+	for (const table of tables) {
+		table.classList.add('rgh-showing-whitespace');
+
+		for (const line of select.all('.blob-code-inner', table)) {
+			showWhiteSpacesOn(line);
```

Comment:
If you add `await true // disable-eslint` here it should give the browser some breathing room to paint frames whenever needed.

This is particularly a problem on the occasional huge page like: https://github.com/Microsoft/TypeScript/blob/master/lib/lib.dom.d.ts

<img width="513" alt="" src="https://user-images.githubusercontent.com/1402241/60575942-03cb3f00-9daf-11e9-9d40-4aa36ff4c988.png">

Reply:
Can you test https://github.com/sindresorhus/refined-github/pull/2073/commits/78be55684163bf1b23457814ee35f7a927afa424 for me and see if it still blocks main thread.

Reply:
Strangely it doesn't work. I think I had tried `while(true){await Promise.resolve()}` in the console and it seemed to work but it doesn't anymore.

Adding this in the loop works but it's crazy slow, of course:

```js
await new Promise(resolve => setTimeout(resolve));
```

Perhaps we should only await it every 100 lines or so

```js
if (i % 100 === 0) {
  await new Promise(resolve => setTimeout(resolve));
}
```